### PR TITLE
Refactor SRTServletRequest to enable dynamic proxies

### PR DESCRIPTION
**Pull request by andymc12**
JAX-RS 3.0 (RESTEasy) injects `HttpServletRequest` instances using dynamic Java proxies that delegate to SRTServletRequest40. This causes problems with the existing web container code because it cannot cast a proxy (that implements the same interfaces as the SRTServletRequest40) to a SRTServletRequest40 or SRTServletRequest - this is because you cannot cast a proxy to a class - only an interface.

The solution I've proposed here is to create interfaces `ISRTServletRequest` and `ISRTServletRequest40` to define the methods needed - then the proxy can implement those interfaces and the casts will succeed.

While the main changes here are in the web container code, there are also some changes to security code and the JAX-RS code.

### CODE REVIEW

**Comment by andymc12** [comment]
#build

**Comment by andymc12** [comment]
#build

File: dev/com.ibm.ws.webcontainer/src/com/ibm/wsspi/webcontainer/util/ServletUtil.java [hunk 1]
```diff
@@ -65,6 +71,7 @@
 	{
 		ServletRequest r = req;
 		while (!(requestClass.isInstance(r)))
+		//while (!(requestClass.isAssignableFrom(r.getClass())))
```
Comment (jhanders34) [inline dev/com.ibm.ws.webcontainer/src/com/ibm/wsspi/webcontainer/util/ServletUtil.java:74]: Any reason this commented out code that you added should stay in?

File: dev/com.ibm.ws.webcontainer.servlet.4.0/src/com/ibm/ws/webcontainer40/srt/ISRTServletRequest40.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+ * Contributors:
+ *      IBM Corporation - initial API and implementation
+ *******************************************************************************/
+package com.ibm.ws.webcontainer40.srt;
+
+import javax.servlet.http.HttpServletMapping;
+
+import com.ibm.ws.webcontainer40.osgi.webapp.WebAppDispatcherContext40;
+
+public interface ISRTServletRequest40 {
```
Comment (jhanders34) [inline dev/com.ibm.ws.webcontainer.servlet.4.0/src/com/ibm/ws/webcontainer40/srt/ISRTServletRequest40.java:16]: Should this interface extend ISRTServletRequest so you don't have to cast to both if you need methods on each of them?

**Review by jhanders34 — CHANGES_REQUESTED** [review]

**Comment by andymc12** [comment]
#build

**Review by jhanders34 — APPROVED** [review]

**Comment by andymc12** [comment]
#build

**Review by pmd1nh — APPROVED** [review]

**Comment by andymc12** [comment]
#libby

**Pull request merged**